testutils/resourcebuilder/machine/v1/nutanix_provider_spec.go

```diff
@@ -0,0 +1,132 @@
+		Image: machinev1.NutanixResourceIdentifier{
+			Type: machinev1.NutanixIdentifierName,
+			Name: ptr.To[string]("rhcos"),
+		},
+		Subnets:        []machinev1.NutanixResourceIdentifier{{Type: machinev1.NutanixIdentifierName, Name: ptr.To[string]("default-net")}},
+		VCPUsPerSocket: int32(1),
+		VCPUSockets:    int32(4),
+		MemorySize:     resource.MustParse(fmt.Sprintf("%dMi", 8096)),
+		Cluster: machinev1.NutanixResourceIdentifier{
+			Type: machinev1.NutanixIdentifierUUID,
+			UUID: ptr.To[string]("pe-uuid"),
+		},
+		SystemDiskSize: resource.MustParse(fmt.Sprintf("%dGi", 120)),
+	}
+
+	if n.withFailureDomains {
+		infra := configv1resourcebuilder.Infrastructure().AsNutanixWithFailureDomains("nutanix-test", nil).Build()
+		failureDomain := infra.Spec.PlatformSpec.Nutanix.FailureDomains[n.failureDomainIndex]
+
+		providerConfig.FailureDomain = &machinev1.NutanixFailureDomainReference{
+			Name: failureDomain.Name,
+		}
+		// update Cluster
+		providerConfig.Cluster = machinev1.NutanixResourceIdentifier{
+			Name: failureDomain.Cluster.Name,
+			UUID: failureDomain.Cluster.UUID,
+		}
+		if failureDomain.Cluster.Type == configv1.NutanixIdentifierName {
+			providerConfig.Cluster.Type = machinev1.NutanixIdentifierName
+		} else if failureDomain.Cluster.Type == configv1.NutanixIdentifierUUID {
+			providerConfig.Cluster.Type = machinev1.NutanixIdentifierUUID
+		}
+
+		// update Subnets
+		providerConfig.Subnets = []machinev1.NutanixResourceIdentifier{}
+
+		for _, fdSubnet := range failureDomain.Subnets {
+			pcSubnet := machinev1.NutanixResourceIdentifier{
+				Name: fdSubnet.Name,
+				UUID: fdSubnet.UUID,
+			}
+			if fdSubnet.Type == configv1.NutanixIdentifierName {
+				pcSubnet.Type = machinev1.NutanixIdentifierName
+			} else if fdSubnet.Type == configv1.NutanixIdentifierUUID {
+				pcSubnet.Type = machinev1.NutanixIdentifierUUID
+			}
+
+			providerConfig.Subnets = append(providerConfig.Subnets, pcSubnet)
+		}
+	}
+
+	return providerConfig
+}
+
+// BuildRawExtension builds a new Nutanix machine config based on the configuration provided.
+func (n *NutanixMachineProviderConfigBuilder) BuildRawExtension() *runtime.RawExtension {
+	providerConfig := n.Build()
+
+	raw, err := json.Marshal(providerConfig)
+	if err != nil {
+		// As we are building the input to json.Marshal, this should never happen.
+		panic(err)
+	}
+
+	return &runtime.RawExtension{
+		Raw: raw,
+	}
+}
+
+// withFailureDomains sets the withFailureDomains field with the input value.
+func (n *NutanixMachineProviderConfigBuilder) WithFailureDomains(withFailureDomains bool) *NutanixMachineProviderConfigBuilder {
+	n.withFailureDomains = withFailureDomains
+	return n
+}
+
+// WithFailureDomainIndex sets the failureDomainIndex field with the input value.
+func (n *NutanixMachineProviderConfigBuilder) WithFailureDomainIndex(index int32) *NutanixMachineProviderConfigBuilder {
+	n.failureDomainIndex = index
+	return n
+}
```

Comment:
This should take a name to a failure domain rather than an index, this should not rely on the infrastructure object since that can be created with custom failure domains

That or it needs to be passed the failure domain from an external infrastructure object, but it definitely should not be defaulting like that

Reply:
Changed to use failureDomainName instead of index, as suggested.

Reply:
Changed to allow to set the failureDomains data externally with the "withFailureDomains()" function.